From 9579e37a197df3f4e824bb1da7990dd14f36ebc6 Mon Sep 17 00:00:00 2001
From: Feng Zhang <feng@radxa.com>
Date: Wed, 15 Jul 2026 17:50:24 +0800
Subject: [PATCH] sunxi: a733: support UFS boot in SPL

The A733 BootROM reports UFS boot with boot source value 0x15.
Map this value to BOOT_DEVICE_UFS so SPL selects the UFS loader.

Propagate errors from SCSI scanning and UFS command submission instead
of continuing after initialization or transfer failures. For SPL UFS
boot, limit the LUN scan range based on the configured raw U-Boot
device number.

These changes provide the code support required for UFS boot in the
A733 SPL.

Signed-off-by: Feng Zhang <feng@radxa.com>
---
 src/arch/arm/include/asm/arch-sunxi/spl.h |  1 +
 src/arch/arm/mach-sunxi/board.c           |  2 ++
 src/common/spl/spl_ufs.c                  |  5 ++++-
 src/drivers/ufs/ufs-uclass.c              | 11 ++++++++---
 4 files changed, 15 insertions(+), 4 deletions(-)

diff --git a/src/arch/arm/include/asm/arch-sunxi/spl.h b/src/arch/arm/include/asm/arch-sunxi/spl.h
index beb0fceb5f3..40a680b73ed 100644
--- a/src/arch/arm/include/asm/arch-sunxi/spl.h
+++ b/src/arch/arm/include/asm/arch-sunxi/spl.h
@@ -17,6 +17,7 @@
 #define SUNXI_BOOTED_FROM_MMC2	2
 #define SUNXI_BOOTED_FROM_SPI	3
 #define SUNXI_BOOTED_FROM_SPI_A733	0x43
+#define SUNXI_BOOTED_FROM_UFS_A733	0x15
 #define SUNXI_BOOTED_FROM_MMC0_HIGH	0x10
 #define SUNXI_BOOTED_FROM_MMC2_HIGH	0x12
 
diff --git a/src/arch/arm/mach-sunxi/board.c b/src/arch/arm/mach-sunxi/board.c
index 0babd509e2e..2a1d719ef58 100644
--- a/src/arch/arm/mach-sunxi/board.c
+++ b/src/arch/arm/mach-sunxi/board.c
@@ -322,6 +322,8 @@ uint32_t sunxi_get_boot_device(void)
 	case SUNXI_BOOTED_FROM_SPI:
 	case SUNXI_BOOTED_FROM_SPI_A733:
 		return BOOT_DEVICE_SPI;
+	case SUNXI_BOOTED_FROM_UFS_A733:
+		return BOOT_DEVICE_UFS;
 	}
 
 	panic("Unknown boot source %d\n", boot_source);
diff --git a/src/common/spl/spl_ufs.c b/src/common/spl/spl_ufs.c
index cef1843f40f..f15530c1f92 100644
--- a/src/common/spl/spl_ufs.c
+++ b/src/common/spl/spl_ufs.c
@@ -30,7 +30,10 @@ static int spl_ufs_load_image(struct spl_image_info *spl_image,
 	int err;
 
 	/* try to recognize storage devices immediately */
-	scsi_scan(false);
+	err = scsi_scan(false);
+	if (err)
+		return err;
+
 	bd = blk_get_devnum_by_uclass_id(UCLASS_SCSI, devnum);
 	if (!bd)
 		return -ENODEV;
diff --git a/src/drivers/ufs/ufs-uclass.c b/src/drivers/ufs/ufs-uclass.c
index 6a51f337e47..e3f9eb66266 100644
--- a/src/drivers/ufs/ufs-uclass.c
+++ b/src/drivers/ufs/ufs-uclass.c
@@ -1660,7 +1660,7 @@ static int ufs_scsi_exec(struct udevice *scsi_dev, struct scsi_cmd *pccb)
 {
 	struct ufs_hba *hba = dev_get_uclass_priv(scsi_dev->parent);
 	u32 upiu_flags;
-	int ocs, result = 0;
+	int ocs, ret, result = 0;
 	u8 scsi_status;
 
 	ufshcd_prepare_req_desc_hdr(hba, &upiu_flags, pccb->dma_dir);
@@ -1669,7 +1669,9 @@ static int ufs_scsi_exec(struct udevice *scsi_dev, struct scsi_cmd *pccb)
 
 	ufshcd_cache_flush(pccb->pdata, pccb->datalen);
 
-	ufshcd_send_command(hba, TASK_TAG);
+	ret = ufshcd_send_command(hba, TASK_TAG);
+	if (ret)
+		return ret;
 
 	ufshcd_cache_invalidate(pccb->pdata, pccb->datalen);
 
@@ -2187,7 +2189,10 @@ int ufshcd_probe(struct udevice *ufs_dev, struct ufs_hba_ops *hba_ops)
 
 	scsi_plat = dev_get_uclass_plat(scsi_dev);
 	scsi_plat->max_id = UFSHCD_MAX_ID;
-	scsi_plat->max_lun = UFS_MAX_LUNS;
+	if (CONFIG_IS_ENABLED(UFS_SUPPORT))
+		scsi_plat->max_lun = CONFIG_SPL_UFS_RAW_U_BOOT_DEVNUM + 1;
+	else
+		scsi_plat->max_lun = UFS_MAX_LUNS;
 	scsi_plat->max_bytes_per_req = UFS_MAX_BYTES;
 
 	hba->dev = ufs_dev;
-- 
2.43.0

